Write down a pagination philosophy (#69719)
Expensify/App: contributingGuides/philosophies/PAGINATION.md

## Why lazy-load data?
The goal of lazy-loading is to decouple the volume of data a user has access to from the app's performance when they use it. This makes it important to ensure that our largest, high-value customers share the same great app experience as a smaller customer.

It makes sense to lazy-load a resource when the volume of that resource is unbounded, or the upper bound of the resource is too costly to load all at once.

## Pagination vs. lazy-loading
As stated above, pagination is a specialized subset of lazy-loading for ordered lists of data. The advantage of structured pagination over a more naïve lazy-loading strategy is that it divides data into discrete chunks such that they can be loaded in batches (and that the order of those batches lends itself to a better product experience).

One example of data that's appropriate for pagination is reportActions (chat messages). We can load the 50 most recent messages, then scroll back and send one network request to get the next batch of 50.

> One example of data that's appropriate to lazy-load but not necessarily ideal for pagination is avatar images. It's too costly to load all avatar images for all users when the app loads, so we only fetch avatar images that a user wants to see on-demand. Another example might be `personalDetails`.

Author: removed the mention of personalDetails!

## How to implement pagination

> 1. Clearly define a deterministic sorting order of the data in question.

Reviewer: NAB bc probably does not need to be pointed out, but are we admitting that we are kind of "without options" philosophy/guidance wise if a list of data is not deterministically sorted? I can only think of one example of course and you know which one it is 😄

Author: I should clarify that "deterministic" doesn't mean that the order can't change if the data changes. It just means that if the same data is sorted, the resultant sort order is _exactly_ the same each time.

To explain with an example, the JS array `.sort` function takes a `compareFunction` that's called with two args, `a` and `b`. The return value of `compareFunction` is:

- negative if `a` should come before `b`
- positive if `a` should come after `b`
- zero if `a` and `b` are considered equal

Our sorting order must never return zero - no two items are allowed to be equal.

Author: I re-wrote this a bit and I think it's clearer now

2. Define a "cursor" - a unique piece of data that encapsulates all the data you need to pinpoint the location of a single item in the list.

>     - Each item in the list _MUST_ be unique. No ties allowed: typically this means your cursor should include an ID.

Reviewer: ties?

>     - Each item in the list _MUST_ be unique. No ties allowed: typically this means your cursor should include an ID.

Reviewer 2: Can you please make the qualifier words (ie. "MUST", "SHOULD", etc.) non-italicized? Just for consistency with the other docs.

Author: done

>     - The cursor _MUST_ be serializable so it can be sent in a network request.

Reviewer: What would an example of a non-serializable cursor be?

>     - The cursor _SHOULD_ include only the minimal fields required to define the location of an item in the list.

Reviewer 2: Another note on consistency with the other docs is that there is usually a "Rules" section of the page, and that's where all the SHOULD/MUST rules go.

I think your explanations and examples are pretty good, but I suggest adding a Rules section to the bottom that clearly lists all the things that should or should not be done.

Author: done

    - _Example:_ For `reportActions`, we use the `created` timestamp, since `reportActions` are generally sorted by order of creation. A better cursor would have been a composite of:

        ```ts
        type ReportActionCursor = {
            /** The datetime when the reportAction was created, in ms precision. */
            created: string;

            /** The actionName is used as a tie-breaker when multiple reportActions are created in the same millisecond (i.e: CREATED actions always come first). */
            actionName: ReportActionName;

            /** The reportActionID is used as a tie-breaker when multiple reportActions of the same type are created in the same millisecond. */
            reportActionID: string;

>         }

Reviewer: Are we currently using a cursor like this anywhere or are we planning to?

Author: I hold this out as a better example, because it includes every piece of data that's used for reportAction sorting.

Author: I do say "A better cursor would have been...", do you still think that's too confusing?

Reviewer: My thoughts on this have not really changed. This section reads like this to me:

- This is our definition of a "cursor".
- We have not put this into practice (no real explanation about why).
- Here's an example of what we should be doing, but are not.

        ```

>     - _Example:_ For a single search snapshot, an integer `sequenceNumber` is used. This is a simple implementation, but generally is not a scalable solution, because it requires that all the data you're paging over needs to be sorted before you can access it by an integer offset and limit.

Reviewer: Maybe better to say some variation of: 

>offset-based approaches require the database to scan and discard all prior rows up to the offset, which becomes slow for large dataset pagination

not sure if that is exactly the problem, but I think it would help to explain the exact problem as we have experienced it and be specific about why `sequenceNumber` bad.

Reviewer: Also, maybe define what is cursor and offset somewhere. I think they are easy to confuse and not exactly the same, but we should make clear the difference and which one we prefer.

Author: updated, thanks for clarifying @marcaaron that's much better. Though I didn't have a super clear explanation for the secondary problem @luacmartins described, so I didn't include it.

Author: updated

>     - _Example:_ For a single search snapshot, an integer `sequenceNumber` is used. This is a simple implementation, but generally is not a scalable solution, because it requires that all the data you're paging over needs to be sorted before you can access it by an integer offset and limit.

Author: I think we made this clearer 👍🏼 

@chuckdries the problem with sequenceNumbers is basically that you need to query everything before the offset in order to grab a single page.

        ```sql
        CREATE TABLE chatMessages (
            id       INTEGER PRIMARY KEY,     -- unique message ID
            created  DATETIME NOT NULL,       -- message creation timestamp
            message  TEXT NOT NULL
        );

        -- Create a sorted covering index for the cursor (sorting by created, tie-breaking with IDs)
        CREATE INDEX idx_chatMessages_created_id
            ON chatMessages (created, id);

        -- Query a single page using your cursor as the "OFFSET"
        SELECT *
        FROM chatMessages
        WHERE (created, id) < ('2025-09-01 10:02:00', 3)
        ORDER BY created DESC, id DESC
        LIMIT 50;
        ```

4. Determine whether your use-case calls for **unidirectional** or **bidirectional** pagination. Generally, **bidirectional** pagination will be useful if:

    - The list live-updates and the user can "jump" to an arbitrary point in the list without loading all the data between the start of the list and that point.
    - The list doesn't live-update (i.e: if a user scrolls back in the list, and while they're looking at older items, they're not receiving newer items as they come in).

>     If you need bidirectional pagination, you'll need to ensure that you craft queries and API endpoints to fetch data from a given cursor in both directions. In the example query above, it could be as simple as switching `<` to `>=` and making the comparator and sort order explicit for each direction.

Reviewer 2: To keep the language consistent in the docs, don't address the user as "you". Instead, reword these like:

> If bidirectional pagination is needed, ensure that queries and API endpoints are crafted to fetch data from a given cursor in both directions.

You did this for the most part in this doc, but there are a couple of spots here (and in point 4 above) where this slipped.

> [!NOTE]
> At the time of writing, [RecyclerListView](https://github.com/Flipkart/recyclerlistview) and [FlashList](https://github.com/Shopify/flash-list) _do not_ support bidirectional pagination.
> React Native's built-in [FlatList](https://reactnative.dev/docs/flatlist) and [legend-list](https://github.com/LegendApp/legend-list) are some examples that _do_ support bidirectional pagination via the `onStartReached` prop.

> 5. Determine whether it's possible for **gaps** to appear in your data. It's non-trivial to list all the ways gaps can appear in a list, but here are a couple of examples, one real and one contrived:

Reviewer: Maybe a terminology section would be useful? Let's define what a "gap" is there.

> 5. Determine whether it's possible for **gaps** to appear in your data. It's non-trivial to list all the ways gaps can appear in a list, but here are a couple of examples, one real and one contrived:

Reviewer: ```suggestion
5. Determine whether it's possible for **gaps** to appear in your data. Here are some examples to help illustrate what a "gap" is and how we might end up with one:
```

    - "Comment linking"

>         1. User has a simple paginated list of integers in ascending order, and it currently contains items 1-50.

Reviewer: ```suggestion
        1. User has a simple paginated list of integers in ascending order, and the frontend has loaded items 1-50.
```

>         2. User jumps to the middle of the list (say for simplicity that they are looking at items 15-35)

Reviewer: ```suggestion
        2. User jumps to the middle of the list and the viewport contains items 15-35
```

>         3. While they're looking at the middle of the list, more than 1 page (50 items) of data is added to the front of the list. Let's say these items are items 50-150.

Author: It doesn't translate to this minimal example, but in practice it happens because other users send messages in the chat report.

>         4. They jump back to the front of the page, and fetch the page at the front, 100-150.

Reviewer: Visual aid could be good here? The language around middle, front, back to the front, fetch at the front etc got confusing for me.

>         4. They jump back to the front of the page, and fetch the page at the front, 100-150.

Reviewer: ```suggestion
        4. They jump to the front of the list and then fetch page 100-150.
```

Reviewer: In lieu of a complete explanation of "comment linking" we could also say "They deep link to a comment in the front" or something like that. It might not be clear what it means to "jump".

            - _Note:_ Fetching "all the data they missed" generally isn't a scalable solution, because it's unbounded, and the unbounded loading of data is the problem pagination seeks to solve.
        5. Now there is a gap between items 51-100 :boom:
    - "Over-eager eviction"
        1. A user scrolls far back in the list. So far that we decide there's too much data for us to handle all at once. Performance degrades, and we decide to start evicting the data at the front that's less-recently viewed. Let's say that the user now has items 50-100 loaded, and we've evicted items 101-200.
        2. New data appears at the front of the list, say items 201-250. We add these items to the list as we normally would if they're looking at the front of the list.
        3. Now there is a gap between items 101-200 (the data we evicted) :boom:

>     If it turns out you _do_ need a strategy for gap detection, here's a high-level summary of how you'd handle it:

Reviewer: Have we thought about what data types we have that will might require gap detection? I think this write up is really great, but mainly curious if we think we will face a similar problem again.

    1. Start keeping track of the pages you've loaded in a dedicated Onyx key for pagination (e.g., `report_123_actions_pages`). This is a _sorted_ list of ranges.
    2. Post-process network responses using the [Pagination middleware](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/Middleware/Pagination.ts) to keep track of the start and end point of the page you loaded in a request, and [merge it with existing pages if it overlaps](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/PaginationUtils.ts#L104).
    3. When rendering your list, use that pages key for your sorting order, and insert "gap markers" between the edges of the pages you've loaded.
    4. When rendering your list, [only render a single continuous chunk](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/PaginationUtils.ts#L166) containing your current "anchor point" (the reportAction you linked to, for example), up until you reach the end of the list in either direction or a gap marker.

>     5. Then when you scroll, you'll hit your gap marker and can make network requests to fill in the gap.

Reviewer: There is a lot of terminology here like "single continuous chunk" and "gap markers". I'm not sure if they need to be explained here or if the Pagination middleware stuff is documenting it better, but this explanation felt almost too generic to be useful. I think we should either simplify it, add more details, or just direct people to the code itself for anyone who needs to understand further.

    More details can be found in [the Pagination middleware](https://github.com/Expensify/App/blob/1a06fa4add10b53a1a9266927d3b08a4ca35d3c4/src/libs/Middleware/Pagination.ts). Efforts were made to generalize this code, but so far it has only been used for reportActions.

## Future improvements

> The following ideals have not (yet) been implemented in Expensify, but we'll hold them out here as aspirational goals for future pagination systems, inspired by WhatsApp.

Reviewer: I like the insights here, but a part of me is 👎 on adding aspirational things to a doc meant to explain the current best practices. Feels like something that should go in a P/S or a design doc and then eventually we can update this. For that reason, I'd suggest deleting everything below this comment.

Reviewer 2: I don't feel too strongly about it. I see your point but I also think this gives _some_ small guidance into the future and that we don't necessarily want to do anything that would really hinder us from going in this direction.

A lot of our original philosophies were very aspirational at the time, like the server pushing data to a client's disk. In the early stages, we were only doing that for reportActions and all other data was loaded with GET requests. Now, we let the server push any and all data to clients via `Onyx.update()`.

Reviewer: We could address that concern by using some strong language and mentioning specifically that we should never move forward with any solution that involves throwing away data on disk since a future solution might rely on it and point to the slack discussion.

My main concern is that we are including specifics about what we might want to build without a clearly defined problem. Putting solutions before problems and getting too attached to specific solutions are typically discouraged at the org level and I think we should model those values here.

The pagination systems we've described so far paginate data from the client to the server. i.e: client loads one page, then when they reach the edge of the page they ask the server for the next page. This goes a long way to decouple the total volume of data a user has access to from their app's performance.

However, the systems we've built and/or described have no data eviction. Once data has been downloaded from the server, it's saved to disk, and because of the design of Onyx all that data is typically kept in RAM in the Onyx cache. If this starts to become a problem ([it has been discussed before](https://expensify.slack.com/archives/C05LX9D6E07/p1755077335799619)), throwing away data on disk is not the best approach. Instead, we should build a two-layer pagination system:

1. Data is loaded into RAM one page at a time.
2. If we reach the edge of the page, we first check the disk for more.
3. Only if the page we requested is not on disk, we request it from the server.
4. If data must be evicted due to slow performance, it is first evicted from RAM.
5. Data should be evicted from disk only if reading data from disk becomes a bottleneck, or we reach device storage limits.

> That way, we can preserve our [Offline Philosophy](https://github.com/Expensify/App/blob/main/contributingGuides/philosophies/OFFLINE.md) and provide a first-class offline UX without compromising performance.

Reviewer 2: FYI these links can be shorted to 
```suggestion
That way, we can preserve our [Offline Philosophy](/contributingGuides/philosophies/OFFLINE.md) and provide a first-class offline UX without compromising performance.
```

Which is better if the repo is ever renamed
